wallet-new/src/Cardano/Wallet/Kernel/Transactions.hs

```diff
@@ -223,6 +234,152 @@ newUnsignedTransaction ActiveWallet{..} options accountId payees = runExceptT $
+                      -> HD.HdAddress
+srcAddressToHdAddress db srcAddress =
+    either (error "Impossible happened: source address doesn't have corresponding HD-address!")
+           id
+           (lookupCardanoAddress db srcAddress)
+
+-- | Submits extrenally-signed transaction to the blockchain.
+-- The result of this function is equal to the result of 'pay' function.
+submitSignedTx :: ActiveWallet
+               -> Tx
+               -> NonEmpty (Address, Signature TxSigData, PublicKey)
+               -> IO (Either PaymentError (Tx, TxMeta))
+submitSignedTx aw@ActiveWallet{..} tx srcAddrsWithProofs = do
+    retrying retryPolicy shouldRetry $ \rs -> do
+        -- STEP 1: create witnesses.
+        -- Since we already received inputs signatures with corresponding derived PKs,
+        -- just form witnesses from them.
+        let witnesses = V.fromList . NonEmpty.toList $ flip NonEmpty.map srcAddrsWithProofs $
+                \(_srcAddr, txSig, derivedPK) -> PkWitness derivedPK txSig
+
+        -- STEP 2: make 'TxAux'.
+        let txAux = TxAux tx witnesses
+
+        -- STEP 3: Compute metadata
+        let txId = hash tx
+        -- Currently it's assumed that all source addresses belong to 'the same/ account,
+        -- so we can just take the first source address to find our 'HdAccountId'.
+        let (firstSrcAddress, _, _) = NonEmpty.head srcAddrsWithProofs
+        snapshot <- liftIO $ getWalletSnapshot walletPassive
+        let (HD.HdAddress (HD.HdAddressId srcAccountId _) _) = srcAddressToHdAddress snapshot firstSrcAddress
+        -- We use `getCreationTimestamp` provided by the `NodeStateAdaptor`
+        -- to compute the createdAt timestamp for `TxMeta`.
+        txMetaCreatedAt_ <- liftIO $ Node.getCreationTimestamp (walletPassive ^. walletNode)
+
+        -- STEP 4: Get available UTxO
+        availableUtxo <- runExceptT $ withExceptT NewTransactionUnknownAccount $ exceptT $
+            currentAvailableUtxo snapshot srcAccountId
+        case availableUtxo of
+            Left _ ->
+                return . Left . PaymentNewTransactionError . NewTransactionUnknownAccount . UnknownHdAccount $
+                    srcAccountId
+            Right utxo -> do
+                let inputsWithCoins = NonEmpty.map (collectInputCoins utxo) $ _txInputs tx
+                -- We have to calculate the sum of input coins.
+                let spentInputCoins = paymentAmount (toaOut . snd <$> inputsWithCoins)
+
+                -- STEP 5: form meta-data.
+                partialMeta <- liftIO $ createNewMeta srcAccountId
+                                                      txId
+                                                      txMetaCreatedAt_
+                                                      inputsWithCoins
+                                                      (_txOutputs tx)
+                                                      True
+                                                      spentInputCoins
```

Comment:
I am not entirely sure this is the right way to go (summoning @kderme here as well). You cannot (I think) simply generate the `TxMeta` for this transaction here, because by the time you are creating the unsigned transaction and the time you actually sign it, a lot of time can have passed (hours if not days). This means that your UTXO might be _completely different_ by the time you call `currentAvailableUtxo`.

I think this requires further thought. Ideally `submitSignedTx` shouldn't do any of this crazyness: it should simply produce a `TxAux` and delegate to the submission layer by calling `newPending`.

I can see _why_ you did that: that's because `newPending` does require you to supply a `partialMeta`. But I think it's not correct to do so here. @kderme , any suggestions?